### PR TITLE
[SPARK-14863][SQL] Cache TreeNode's hashCode by default

### PR DESCRIPTION
**Pull request by JoshRosen**
Caching TreeNode's `hashCode` can lead to orders-of-magnitude performance improvement in certain optimizer rules when operating on huge/complex schemas.

### CODE REVIEW

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 1]
```diff
@@ -71,7 +71,9 @@ object CurrentOrigin {
   }
 }
 
+// scalastyle:off
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: why is this needed?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: Because we override `hashCode` but not `equals`, which breaks the style check.

An alternative solution is to override `equals` too, I don't have a strong preference to it.

Comment (JoshRosen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: I tried overriding equals but it seems to break things pretty badly because it seemingly overrides the case-class equals which would be auto-generated in subclasses.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: would be better to turn off style check for this specific rule so it is more clear since it is not obvious why this needs to be excluded looking at this line.

Comment (JoshRosen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: I actually tried this by assigning a custom name to the rule and it ended up breaking the Scalastyle error message output for this rule. If that had not happened, I would have used a more specific exclude here.
